File: server-python/hall_server/room_service.py
"""游戏服上报接口（:9002）+ 向游戏服发起内部调用。

对应 `server/hall_server/room_service.ts`。

`/register_gs` 是游戏服每 5 秒心跳上报的入口；`create_room` / `enter_room` 是大厅服
主动调游戏服内部 HTTP 接口的两个函数，签名拼接顺序必须与 `game_server/http_service.py`
逐字一致——改任一侧都是"大厅服调不通游戏服"。
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any

from aiohttp import web

from utils import crypto, db, http

#: `config` 由 `create_routes()` 在开始监听前赋值，所有请求处理里必然已经就绪。
_config: dict[str, Any] | None = None
_hall_ip: str | None = None

#: 房间号 -> 游戏服地址（`ip:port`）。
rooms: dict[str, str] = {}
#: 游戏服地址（`clientip:clientport`）-> 上报信息。
server_map: dict[str, dict[str, Any]] = {}
#: userId -> 房间号。
room_id_of_users: dict[str, str] = {}


@dataclass
class EnterInfo:
    """`create_room` / `enter_room` 交给调用方的进房信息：`token` 由游戏服 `/enter_room` 下发。"""

    ip: str | None
    port: str | None
    token: str


def _port_of(info: dict[str, Any]) -> int | None:
    """把上报里的 `httpPort`（query 字符串）转成 aiohttp 需要的整数端口。

    Node 版把它原样交给 `http.get` 的 `options.port`，Node 接受字符串；
    Python 侧必须显式转一次，转不出来就当作"没给端口"。
    """
    value = info.get("httpPort")
    if value is None:
        return None
    try:
        return int(str(value))
    except ValueError:
        return None


def create_routes(config: dict[str, Any]) -> web.Application:
    """建大厅服的上报服务应用。"""
    global _config
    _config = config

    app = web.Application()

    @web.middleware
    async def cors(request: web.Request, handler: Any) -> web.StreamResponse:
        response = await handler(request)
        response.headers["Access-Control-Allow-Origin"] = "*"
        response.headers["Access-Control-Allow-Headers"] = "X-Requested-With"
        response.headers["Access-Control-Allow-Methods"] = "PUT,POST,GET,DELETE,OPTIONS"
        response.headers["X-Powered-By"] = " 3.2.1"
        if "Content-Type" not in response.headers:
            response.headers["Content-Type"] = "application/json;charset=utf-8"
        return response

    app.middlewares.append(cors)

    async def on_register_gs(request: web.Request) -> web.Response:
        ip = http.request_ip(request)
        clientip = http.query_string(request, "clientip")
        clientport = http.query_string(request, "clientport")
        http_port = http.query_string(request, "httpPort")
        load = http.query_string(request, "load")
        server_id = str(clientip) + ":" + str(clientport)

        if server_id in server_map:
            info = server_map[server_id]
            if (
                info.get("clientport") != clientport
                or info.get("httpPort") != http_port
                or info.get("ip") != ip
            ):
                print("duplicate gsid:" + server_id + ",addr:" + ip + "(" + str(http_port) + ")")
                return http.send(1, "duplicate gsid:" + server_id)
            info["load"] = load
            return http.send(0, "ok", {"ip": ip})

        server_map[server_id] = {
            "ip": ip,
            "id": server_id,
            # 上面的 id 已经用 `clientip + ":" + clientport` 算过；这里显式转字符串，
            # 与老代码把值直接放进对象的隐式转字符串结果相同（缺参数时一样是 "None"）。
            "clientip": str(clientip),
            "clientport": str(clientport),
            "httpPort": http_port,
            "load": load,
        }
        print(
            "game server registered.\n\tid:"
            + server_id
            + "\n\taddr:"
            + ip
            + "\n\thttp port:"
            + str(http_port)
            + "\n\tsocket clientport:"
            + str(clientport)
        )

        # 原实现是"先把 HTTP 回包发出去，再异步去拉 /get_server_info"。
        # 这里保持同样的时序：把那次调用丢到后台任务里，不等它就把 ok 回给游戏服
        # （否则游戏服每 5 秒的心跳都要多等一个来回，对方不可达时还会被拖住）。
        reqdata = {
            "serverid": server_id,
            "sign": crypto.md5(server_id + _require_config()["ROOM_PRI_KEY"]),
        }
        _spawn(_fetch_server_info(ip, _port_of(server_map[server_id]), reqdata))
        return http.send(0, "ok", {"ip": ip})

    app.router.add_get("/register_gs", on_register_gs)
    return app


async def _fetch_server_info(ip: str, port: int | None, reqdata: dict[str, Any]) -> None:
    """拉一次游戏服的 `/get_server_info`（原实现里的异步回调体）。"""
    result = await http.get(ip, port, "/get_server_info", reqdata)
    if not result.ok:
        return
    data = result.data
    if data.get("errcode") == 0:
        user_room_info = data.get("userroominfo") or []
        # 原实现只是把数组两两取出来（局部变量随后被丢弃），这里如实保留这个循环。
        for i in range(0, len(user_room_info), 2):
            _ = user_room_info[i]
            _ = user_room_info[i + 1] if i + 1 < len(user_room_info) else None
    else:
        print(data.get("errmsg"))


#: 后台任务的强引用集合（不保留引用的话任务可能被 GC 掉）。
_background_tasks: set[Any] = set()


def _spawn(coroutine: Any) -> None:
    task = asyncio.get_running_loop().create_task(coroutine)
    _background_tasks.add(task)
    task.add_done_callback(_background_tasks.discard)


def choose_server() -> dict[str, Any] | None:
    """挑一台负载最低的游戏服（`load` 按**字符串**比较，与原实现一致）。"""
    serverinfo: dict[str, Any] | None = None
    for info in server_map.values():
        new_load = info.get("load")
        if serverinfo is None:
            serverinfo = info
        else:
            old_load = serverinfo.get("load")
            if old_load is not None and new_load is not None and old_load > new_load:
                serverinfo = info
    return serverinfo


def _require_config() -> dict[str, Any]:
    if _config is None:
        raise RuntimeError("room_service.create_routes() 尚未调用")
    return _config


async def create_room(
    account: str | None,
    user_id: Any,
    room_conf: Any,
) -> tuple[int, str | None]:
    """请求游戏服建房。

    :return: `(errcode, roomId)`；101 没有可用游戏服、102 调不通、103 查不到房卡。
    """
    serverinfo = choose_server()
    if serverinfo is None:
        return 101, None

    data = await db.get_gems(account)
    if data is None:
        return 103, None

    reqdata: dict[str, Any] = {
        "userid": user_id,
        "gems": data.get("gems"),
        # 客户端传来的建房配置：它本来就是 query 里的一个 JSON 字符串，按老代码原样带上。
        "conf": room_conf,
    }
    # 签名拼接顺序与游戏服 http_service.py 完全一致：userId + conf + gems + ROOM_PRI_KEY。
    reqdata["sign"] = crypto.md5(
        str(user_id) + str(room_conf) + str(data.get("gems")) + _require_config()["ROOM_PRI_KEY"]
    )

    result = await http.get(
        str(serverinfo.get("ip")), _port_of(serverinfo), "/create_room", reqdata
    )
    if not result.ok:
        return 102, None

    if result.data.get("errcode") == 0:
        return 0, result.data.get("roomid")
    return int(result.data.get("errcode") or 0), None


async def enter_room(
    user_id: Any,
    name: str | None,
    room_id: str | None,
) -> tuple[int, EnterInfo | None]:
    """请求游戏服进房，返回客户端连游戏服需要的 `(ip, port, token)`。"""
    reqdata: dict[str, Any] = {
        "userid": user_id,
        "name": name,
        "roomid": room_id,
    }
    # 签名拼接顺序与游戏服 http_service.py 完全一致：userId + name + roomId + ROOM_PRI_KEY。
    reqdata["sign"] = crypto.md5(
        str(user_id) + str(name) + str(room_id) + _require_config()["ROOM_PRI_KEY"]
    )

    async def check_room_is_running(serverinfo: dict[str, Any]) -> bool:
        # 与游戏服 /is_room_runing 的校验一致：md5(roomId + ROOM_PRI_KEY)。
        sign = crypto.md5(str(room_id) + _require_config()["ROOM_PRI_KEY"])
        result = await http.get(
            str(serverinfo.get("ip")),
            _port_of(serverinfo),
            "/is_room_runing",
            {"roomid": room_id, "sign": sign},
        )
        if not result.ok:
            # 网络失败等价于"没在跑"。
            return False
        return result.data.get("errcode") == 0 and result.data.get("runing") is True

    async def enter_room_req(serverinfo: dict[str, Any]) -> tuple[int, EnterInfo | None]:
        result = await http.get(
            str(serverinfo.get("ip")), _port_of(serverinfo), "/enter_room", reqdata
        )
        if not result.ok:
            return -1, None
        data = result.data
        print(data)
        if data.get("errcode") == 0:
            await db.set_room_id_of_user(user_id, room_id)
            return 0, EnterInfo(
                ip=serverinfo.get("clientip"),
                port=serverinfo.get("clientport"),
                token=data.get("token"),
            )
        print(data.get("errmsg"))
        return int(data.get("errcode") or 0), None

    ok, ip, port = await db.get_room_addr(room_id)
    if ok:
        server_id = str(ip) + ":" + str(port)
        serverinfo = server_map.get(server_id)
        if serverinfo is not None:
            if await check_room_is_running(serverinfo):
                return await enter_room_req(serverinfo)
            chosen = choose_server()
            if chosen is not None:
                return await enter_room_req(chosen)
            return -1, None
        chosen = choose_server()
        if chosen is not None:
            return await enter_room_req(chosen)
        return -1, None
    return -2, None

File: server/hall_server/room_service.ts
import express from "express";
import type { Request, Response } from "express";

import * as crypto from "../utils/crypto";
import * as db from "../utils/db";
import * as http from "../utils/http";

import type { HallServerConfig } from "../types/config";

var app = express();

var hallIp: string | null = null;
// `config` 由 `start($config)` 在开始监听前赋值，所有请求处理里必然已经就绪；
// 下面的 `config!` 只是把这一点写出来，不改变任何运行时行为。
var config: HallServerConfig | null = null;
var rooms: Record<string, string> = {};
var serverMap: Record<string, ServerInfoInput> = {};
var roomIdOfUsers: Record<string, string> = {};

/**
 * 游戏服通过 `/register_gs` 上报并在本文件里登记的一份信息（字段拼写与迁移前一致）。
 *
 * `clientip` / `clientport` 是 `/enter_room` 成功后回给客户端的连接地址，登记时必有，
 * 所以标成必填——否则 `enterInfo` 里会出现 `undefined`，而 JSON 序列化会把该字段整个丢掉，
 * 那才是行为改变。`clientport` / `httpPort` / `load` 都直接来自 query string，因此是字符串；
 * `chooseServer` 里的 `load` 比较也照旧按字符串比，不改成数值。
 */
interface ServerInfo {
	ip?: string;
	id?: string;
	clientip: string;
	clientport: string;
	httpPort?: string | number;
	load?: string;
}

/** 新登记一台游戏服、并回应游戏服心跳时用的那份信息（每次都从 query 重建）。 */
type ServerInfoInput = ServerInfo;

/** `createRoom` / `enterRoom` 交给调用方的进房信息：`token` 由游戏服 `/enter_room` 下发。 */
export interface EnterInfo {
	ip?: string;
	port?: string;
	token: string;
}

/** 客户端传来的 userId：`t_users.userid` 是 INT，但老代码一律按字符串拼签名与 query。 */
type UserId = string | number;

/** 大厅服调游戏服那四个内部接口返回体里本文件用到的字段。 */
interface ShardResponse {
	errcode: number;
	errmsg: unknown;
	roomid: string;
	token: string;
	runing: boolean;
	userroominfo: unknown;
}

//设置跨域访问
app.all('*', function(req: Request, res: Response, next) {
	res.header("Access-Control-Allow-Origin", "*");
	res.header("Access-Control-Allow-Headers", "X-Requested-With");
	res.header("Access-Control-Allow-Methods","PUT,POST,GET,DELETE,OPTIONS");
	res.header("X-Powered-By",' 3.2.1');
	res.header("Content-Type", "application/json;charset=utf-8");
	next();
});

app.get('/register_gs',function(req: Request,res: Response){
	
	// @types/express 把 req.ip 标成 string | undefined（trust proxy 的边界情况）；
	// 这里与迁移前一样直接按字符串使用，不加存在性判断，行为逐字不变。
	var ip = req.ip as string;
	var clientip = http.queryString(req,"clientip");
	var clientport = http.queryString(req,"clientport");
	var httpPort = http.queryString(req,"httpPort");
	var load = http.queryString(req,"load");
	var id = clientip + ":" + clientport;

	if(serverMap[id]){
		var info = serverMap[id];
		if(info.clientport != clientport
			|| info.httpPort != httpPort
			|| info.ip != ip
		){
			console.log("duplicate gsid:" + id + ",addr:" + ip + "(" + httpPort + ")");
			http.send(res,1,"duplicate gsid:" + id);
			return;
		}
		info.load = load;
		http.send(res,0,"ok",{ip:ip});
		return;
	}
	var newInfo: ServerInfoInput = {
		ip:ip,
		id:id,
		// 上面的 id 已经用 `clientip + ":" + clientport` 算过；这里显式 String(...)，
		// 与老代码把值直接放进对象的隐式转字符串结果相同（缺参数时一样是 "undefined"）。
		clientip:String(clientip),
		clientport:String(clientport),
		httpPort:httpPort,
		load:load
	};
	serverMap[id] = newInfo;
	http.send(res,0,"ok",{ip:ip});
	console.log("game server registered.\n\tid:" + id + "\n\taddr:" + ip + "\n\thttp port:" + httpPort + "\n\tsocket clientport:" + clientport);

	var reqdata: http.QueryParams = {
		serverid:id,
		sign:crypto.md5(id+config!.ROOM_PRI_KEY)
	};
	//获取服务器信息
	http.get(ip,httpPort as number | undefined,"/get_server_info",reqdata,function(result){
		if(!result.ok){
			// 原来是 `ret && data.errcode == 0` 才做处理；网络失败时 ret 为 false，什么都不做。
			return;
		}
		// 返回体来自游戏服 `http_service.js`，字段是我们的内部约定；断言只是把 JSON 的
		// unknown 收窄回迁移前的读法，不做任何运行时判断（原来也没有）。
		var data = result.data as unknown as ShardResponse;
		if(data.errcode == 0){
			var userroominfo = data.userroominfo as unknown[];
			for(var i = 0; i < userroominfo.length; i += 2){
				var userId = userroominfo[i];
				var roomId = userroominfo[i+1];
			}
		}
		else{
			console.log(data.errmsg);
		}
	});
});

function chooseServer(): ServerInfo | null {
	var serverinfo: ServerInfo | null = null;
	for(var s in serverMap){
		var info = serverMap[s];
		// 两次读同一个字段；取到局部变量是为了让类型检查器知道它们已经存在
		// （取值与比较方式都没变，`load` 仍按字符串比较）。
		var newLoad = info.load;
		if(serverinfo == null){
			serverinfo = info;			
		}
		else{
			var oldLoad = serverinfo.load;
			if(oldLoad != null && newLoad != null && oldLoad > newLoad){
				serverinfo = info;
			}
		}
	}	
	return serverinfo;
}

export function createRoom(account: string | undefined,userId: UserId,roomConf: string | number | boolean | null | undefined,fnCallback: (errcode: number,roomId: string | null) => void){
	const serverinfo = chooseServer();
	if(serverinfo == null){
		fnCallback(101,null);
		return;
	}
	db.get_gems(account,function(data: { gems: number } | null){
		if(data != null){
			//2、请求创建房间
			var reqdata: http.QueryParams = {
				userid:userId,
				gems:data.gems,
				// 客户端传来的建房配置：它本来就是 query 里的一个 JSON 字符串，这里按老代码原样带上。
				conf:roomConf as string
			}
			// 签名拼接顺序与游戏服 http_service.js 完全一致：userId + conf + gems + ROOM_PRI_KEY。
			// 这里显式写成 String(...)，与老代码的隐式转字符串是同一个结果。
			reqdata.sign = crypto.md5(String(userId) + String(roomConf) + String(data.gems) + config!.ROOM_PRI_KEY);
			http.get(serverinfo.ip as string,serverinfo.httpPort as number | undefined,"/create_room",reqdata,function(result){
				//console.log(data);
				if(!result.ok){
					// 原来是 `ret` 为真才走成功分支；网络失败时 ret 为 false，直接回 102。
					fnCallback(102,null);
					return;
				}
				// 返回体来自游戏服 /create_room；断言只是收窄 JSON 的 unknown，行为不变。
				var data = result.data as unknown as ShardResponse;
				if(data.errcode == 0){
					fnCallback(0,data.roomid);
				}
				else{
					fnCallback(data.errcode,null);		
				}
			});	
		}
		else{
			fnCallback(103,null);
		}
	});
}

export function enterRoom(userId: UserId,name: string | null,roomId: string | null | undefined,fnCallback: (errcode: number,enterInfo: EnterInfo | null) => void){
	var reqdata: http.QueryParams = {
		userid:userId,
		name:name,
		roomid:roomId
	}
	// 签名拼接顺序与游戏服 http_service.js 完全一致：userId + name + roomId + ROOM_PRI_KEY。
	// 显式的 String(...) 与老代码的隐式转字符串结果相同（null 一样变成 "null"）。
	reqdata.sign = crypto.md5(String(userId) + String(name) + String(roomId) + config!.ROOM_PRI_KEY);

	var checkRoomIsRuning = function(serverinfo: ServerInfo,roomId: string | null | undefined,callback: (isRuning: boolean) => void){
		// 与游戏服 /is_room_runing 的校验一致：md5(roomId + ROOM_PRI_KEY)。
		var sign = crypto.md5(String(roomId) + config!.ROOM_PRI_KEY);
		http.get(serverinfo.ip as string,serverinfo.httpPort as number | undefined,"/is_room_runing",{roomid:roomId,sign:sign},function(result){
			if(!result.ok){
				// 原来是 `ret` 为假的分支；网络失败等价于"没在跑"。
				callback(false);
				return;
			}
			// 返回体来自游戏服 /is_room_runing；断言只是收窄 JSON 的 unknown，行为不变。
			var data = result.data as unknown as ShardResponse;
			if(data.errcode == 0 && data.runing == true){
				callback(true);
			}
			else{
				callback(false);
			}
		});
	}

	var enterRoomReq = function(serverinfo: ServerInfo){
		http.get(serverinfo.ip as string,serverinfo.httpPort as number | undefined,"/enter_room",reqdata,function(result){
			if(!result.ok){
				// 原来是 `ret` 为假的分支。
				fnCallback(-1,null);
				return;
			}
			// 返回体来自游戏服 /enter_room；断言只是收窄 JSON 的 unknown，行为不变。
			var data = result.data as unknown as ShardResponse;
			console.log(data);
			if(data.errcode == 0){
				db.set_room_id_of_user(userId,roomId,function(ret: boolean){
					fnCallback(0,{
						ip:serverinfo.clientip,
						port:serverinfo.clientport,
						token:data.token
					});
				});
			}
			else{
				console.log(data.errmsg);
				fnCallback(data.errcode,null);
			}
		});
	}

	var chooseServerAndEnter = function(serverinfo: ServerInfo | null){
		serverinfo = chooseServer();
		if(serverinfo != null){
			enterRoomReq(serverinfo);
		}
		else{
			fnCallback(-1,null);					
		}
	}

	db.get_room_addr(roomId,function(ret: boolean,ip: string | null,port: number | null){
		if(ret){
			var id = ip + ":" + port;
			var serverinfo = serverMap[id];
			if(serverinfo != null){
				checkRoomIsRuning(serverinfo,roomId,function(isRuning: boolean){
					if(isRuning){
						enterRoomReq(serverinfo);
					}
					else{
						chooseServerAndEnter(serverinfo);
					}
				});
			}
			else{
				chooseServerAndEnter(serverinfo);
			}
		}
		else{
			fnCallback(-2,null);
		}
	});
}

export function start($config: HallServerConfig){
	config = $config;
	// 返回 http.Server，交给 app.js 汇总成启动横幅（见 utils/startup.js）
	return app.listen(config.ROOM_PORT,config.FOR_ROOM_IP);
}
